File: src/main/java/pl/allegro/experiments/chi/chiserver/infrastructure/experiments/FileBasedExperimentsRepository.java
package pl.allegro.experiments.chi.chiserver.infrastructure.experiments;

import com.google.common.collect.Sets;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ReadOnlyExperimentsRepository;
import pl.allegro.experiments.chi.chiserver.infrastructure.InMemoryExperimentsRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class FileBasedExperimentsRepository implements ReadOnlyExperimentsRepository {
    private static final Logger logger = LoggerFactory.getLogger(FileBasedExperimentsRepository.class);
    private final DataLoader dataLoader;
    private final Gson jsonConverter;
    private final InMemoryExperimentsRepository inMemoryExperimentsRepository;
    private String jsonUrl;

    public FileBasedExperimentsRepository(
            String jsonUrl,
            DataLoader dataLoader,
            Gson jsonConverter,
            List<Experiment> initialState) {
        if (initialState == null) {
            initialState = new ArrayList<>();
        }
        this.dataLoader = dataLoader;
        this.jsonConverter = jsonConverter;
        this.inMemoryExperimentsRepository = new InMemoryExperimentsRepository(initialState);
        this.jsonUrl = jsonUrl;
        secureRefresh();
    }


    public void secureRefresh() {
        try {
            refresh();
        } catch (Exception e) {
            logger.error("Error while loading experiments file.", e);
        }
    }

    private void refresh() {
        String data = dataLoader.load(jsonUrl);

        if (data.isEmpty()) {
            logger.error("refresh failed, dataLoader has returned empty String");
            return;
        }

        List<Experiment> freshExperiments;
        try {
            freshExperiments = jsonConverter.fromJson(data, (new TypeToken<List<Experiment>>() {}).getType());
        } catch (IllegalArgumentException e) {
            logger.error("refresh failed, malformed experiments definition in JSON: " + e.getMessage());
            return;
        }
        Set<String> currentExperimentIds = inMemoryExperimentsRepository.experimentIds();
        Set<String> freshExperimentsIds = freshExperiments.stream()
                .map(Experiment::getId)
                .collect(Collectors.toSet());
        Set<String> removedExperiments = Sets.difference(currentExperimentIds, freshExperimentsIds);
        freshExperiments.forEach(inMemoryExperimentsRepository::save);
        removedExperiments.forEach(inMemoryExperimentsRepository::remove);
        logger.debug(freshExperiments.size() + " experiment(s) successfully loaded");
    }

    @Override
    public Experiment getExperiment(String id) {
        return inMemoryExperimentsRepository.getExperiment(id);
    }

    @Override
    public List<Experiment> getAll() {
        return inMemoryExperimentsRepository.getAll();
    }

    @Override
    public List<Experiment> assignable() {
        return inMemoryExperimentsRepository.assignable();
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/infrastructure/experiments/FileBasedExperimentsRepository.kt
package pl.allegro.experiments.chi.chiserver.infrastructure.experiments

import com.github.salomonbrys.kotson.fromJson
import org.slf4j.LoggerFactory
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentId
import pl.allegro.experiments.chi.chiserver.domain.experiments.ReadOnlyExperimentsRepository
import pl.allegro.experiments.chi.chiserver.infrastructure.InMemoryExperimentsRepository
import pl.allegro.experiments.chi.chiserver.infrastructure.JsonConverter
import javax.naming.OperationNotSupportedException

open class FileBasedExperimentsRepository(jsonUrl: String,
                                     private val dataLoader: (String) -> String,
                                     private val jsonConverter: JsonConverter,
                                     initialState: List<Experiment> = emptyList()) : ReadOnlyExperimentsRepository {

    private val inMemoryRepository: InMemoryExperimentsRepository = InMemoryExperimentsRepository(initialState)
    private var jsonUrl: String = jsonUrl

    companion object {
        private val logger = LoggerFactory.getLogger(FileBasedExperimentsRepository::class.java)
    }

    init {
        secureRefresh()
    }

    fun secureRefresh() {
        try {
            refresh()
        } catch (e: Exception) {
            logger.error("Error while loading experiments file.", e)
        }
    }

    private fun refresh() {
        val data = dataLoader.invoke(jsonUrl)

        if (data.isBlank()) {
            logger.error("refresh failed, dataLoader has returned empty String")
            return
        }

        val freshExperiments: List<Experiment>
        try {
            freshExperiments = jsonConverter.fromJson(data)
        } catch (e: IllegalArgumentException) {
            logger.error("refresh failed, malformed experiments definition in JSON: ${e.message}", e)
            return
        }

        val currentExperimentsIds = inMemoryRepository.experimentIds()
        val freshExperimentsIds = freshExperiments.map { it.id }.toSet()

        val removedExperiments = currentExperimentsIds.minus(freshExperimentsIds)

        freshExperiments.forEach { inMemoryRepository.save(it) }
        removedExperiments.forEach { inMemoryRepository.remove(it) }

        logger.debug("{} experiment(s) successfully loaded", freshExperiments.size)
    }

    override fun getExperiment(id: ExperimentId): Experiment? = inMemoryRepository.getExperiment(id)

    override fun getAll() : List<Experiment> = inMemoryRepository.getAll()

    override val overridable: List<Experiment>
        get() = inMemoryRepository.overridable
}
